[flang] Add debug information for module variables.

abidh opened the pull request:
This PR add debug info for module variables. The module variables are added as global variables but their scope is set to module instead of compile unit. The scope of function declared inside a module is also set accordingly.

After this patch, a module variable could be evaluated in the GDB as `p helper::gli` where helper is name of the module and gli is the name of the variable. A future patch will add the import module functionality which will remove the need to prefix the name with helper::.

The line number where is module is declared is a best guess at the moment as this information is not part of the GlobalOp.

abidh committed:
[flang] Add initial implementation of module variables.

This PR add supports for module variables and function. The module
variables are added as global variables but their scope is set to
module instead of compile unit. The scope of function declared inside
a module is also set accordingly.

After this patch, a module variable could be evaluated in the GDB as
`p helper::gli` where helper is name of the module and gli is the name
of the variable. A future patch will add the import module functionality
which will remove the need to prefix the name with helper::.

The line number where is module is declared is a best guess at the
moment as this information is not part of the GlobalOp.

kiranchandramohan commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -1030,10 +1030,15 @@ LogicalResult ModuleTranslation::convertGlobals() {
       llvm::DIGlobalVariable *diGlobalVar = diGlobalExpr->getVariable();
       var->addDebugInfo(diGlobalExpr);
 
+      // For fortran, the scope hierarchy can be
+      // variable -> module -> compile unit
+      llvm::DIScope *scope = diGlobalVar->getScope();
+      if (llvm::DIModule *mod = dyn_cast_if_present<llvm::DIModule>(scope))
+        scope = mod->getScope();
+
```
Changes to MLIR should go in as a separate patch. And a test is required for the MLIR changes in I guess `mlir/test/Target/LLVMIR/llvmir-debug.mlir`.

abidh replied:
Thanks for letting me know. I have now opened #91604 with this change and a testcase. I will remove this bit from this PR.

kiranchandramohan commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -48,10 +49,87 @@ class AddDebugInfoPass : public fir::impl::AddDebugInfoBase<AddDebugInfoPass> {
+
 } // namespace
 
+// The `module` does not have a first class representation in the `FIR`. We
+// extract information about it from the name of the identifiers and keep a
+// map to avoid duplication.
```
Would this mean that if a module variable or subroutine is not used we will not have the information that the module is used via a USE statement?

abidh replied:
In debugger, one should be able to evaluate module variables using module::var syntax even in functions where the module is not used (using `use` keyword). For functions where it is used, evaluation without module:: should work too (once we have implemented the imported module functionality). 

I also tried multi file examples where module is defined in one file and used in others. In such cases, compilers (e.g. gfortran, classic flang) generate a module entry in both the compile units. To differentiate, they set decl attribute to true in files where it is just used. I have tried to follow the similar approach. I found that GlobalOp::hasInitializationBody() gave me this information. It was true in compile unit where the module was actually defined and false where it was just used.

tblah commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -91,6 +169,10 @@ void AddDebugInfoPass::runOnOperation() {
       llvm::dwarf::getLanguage("DW_LANG_Fortran95"), fileAttr, producer,
       isOptimized, debugLevel);
 
+  module.walk([&](fir::GlobalOp globalOp) {
```
`mlir::Operation::walk` will visit every descendant operation. I think `fir::GlobalOp` can only be nested directly inside of the `MoudleOp` so it would be much more efficient to have a for loop over the direct descendants of the module.

abidh replied:
I have changed it to
`for (auto globalOp : module.getOps<fir::GlobalOp>())`

Please let me know if that was not what you had in mind.

tblah replied:
Yes that's what I meant. Thank you!

abidh committed:
Handled review comments.

Following changes were done.

1. Remove MLIR changes as they have been moved to separate PR.

2. Replace std::map with llvm::StringMap

3. Use a direct for loop instead of module.walk.

abidh commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -48,10 +49,87 @@ class AddDebugInfoPass : public fir::impl::AddDebugInfoBase<AddDebugInfoPass> {
   void runOnOperation() override;
+
+private:
+  std::map<std::string, mlir::LLVM::DIModuleAttr> moduleMap;
```
Done.

tblah commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -34,6 +34,7 @@
 #include "llvm/Support/FileSystem.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
+#include <map>
```
nit: this is no longer needed

abidh replied:
Thanks for catching this. Removed.

tblah reviewed (approved):
This looks good to me, thanks!

abidh committed:
Remove a redundant include file.

kiranchandramohan commented on flang/test/Transforms/debug-module-1.f90:
```diff
@@ -0,0 +1,39 @@
+
+! CHECK-DAG: loc(fused<#[[GLR]]>[#[[LOC1]]])
+! CHECK-DAG: loc(fused<#[[GLI]]>[#[[LOC2]]])
+! CHECK-DAG: loc(fused<#[[TEST]]>[#[[LOC3]]])
```
Nit: end of line.

abidh replied:
Fixed.

kiranchandramohan reviewed (commented):
The patch adds debug info for other kind of globals (besides module variables) like for eg. `x` in the following case. Is that the intention? If so a test would be good.
```
subroutine abc
  integer :: x = 2
  print *, x
end subroutine
```
Commonblocks also are globals but I did not see debug for them. 

I am guessing the following points will come in future patches.
-> Renamed module variables using `use` statement.
-> Private module variables. Can they be distinguished in DWARF?
-> Submodules

kiranchandramohan commented:
There is a plan for a pass (https://github.com/llvm/llvm-project/pull/73829) that would convert some constants to globals. Would it be OK to mark this as a Global Variable?

abidh commented:
> The patch adds debug info for other kind of globals (besides module variables) like for eg. `x` in the following case. Is that the intention? If so a test would be good.
> 
> ```
> subroutine abc
>   integer :: x = 2
>   print *, x
> end subroutine
> ```
> 
This was not intentional. The variable with save attribute should be added as global but with scope limited to the subroutine. I will add a check to skip them in this patch.

On a related note, while processing `DeclareOp` of such variables, are you aware of anything that will tell me that these variables have `save` attribute? I ask because I noticed that they are being added as local variable too.

> Commonblocks also are globals but I did not see debug for them.
> 

Yes, common blocks are not supported yet.

> I am guessing the following points will come in future patches. -> Renamed module variables using `use` statement. -> Private module variables. Can they be distinguished in DWARF? -> Submodules

Private module variable should work with this patch. I was not aware of the renaming by the `use` statement syntax. I will add it in my todo list. `DWARF` has concept of `DW_TAG_imported_declaration` for such cases. I will have to see if we have enough information to generate it.

abidh commented:
> There is a plan for a pass (#73829) that would convert some constants to globals. Would it be OK to mark this as a Global Variable?

Currently, we add variable only if deconstructed name kind is `fir::NameUniquer::NameKind::VARIABLE`. I think that should filter out the generated globals.

abidh committed:
Addressed review comments.

1. Skip variables if module list is empty.

2. Add a missing line at the end of a file.

jeanPerier commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
+    return;
+
+  scope = getOrCreateModuleAttr(result.second.modules[0], fileAttr, scope,
+                                line - 1, !globalOp.hasInitializationBody());
```
You should use `globalOp.isInitialized()` here and below. In some cases, the initial value is defined as an attribute, not with operations inside the body.

abidh replied:
Thanks for you suggestions. Changed now.

abidh committed:
Process variable only if full debug info is requested.

abidh committed:
Replace 'hasInitializationBody' with 'isInitialized'.

This address one of the review comments.

kiranchandramohan commented:
> On a related note, while processing `DeclareOp` of such variables, are you aware of anything that will tell me that these variables have `save` attribute? I ask because I noticed that they are being added as local variable too.

I am not aware.

kiranchandramohan commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
+  fir::DebugTypeGenerator typeGen(module);
+  mlir::OpBuilder builder(context);
+
+  auto result = fir::NameUniquer::deconstruct(globalOp.getSymName());
```
Nit: Could you spell auto?

abidh replied:
Done. Although I noticed that auto has been used in most places with `fir::NameUniquer::deconstruct`.

kiranchandramohan commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
+    mlir::LLVM::DIScopeAttr scope, unsigned line, bool decl) {
+  mlir::MLIRContext *context = &getContext();
+  mlir::LLVM::DIModuleAttr modAttr;
+  if (auto iter{moduleMap.find(name)}; iter != moduleMap.end())
+    modAttr = iter->getValue();
```
Nit: Braces here to match the else.

abidh replied:
Done.

kiranchandramohan commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
+  else {
+    modAttr = mlir::LLVM::DIModuleAttr::get(
+        context, fileAttr, scope, mlir::StringAttr::get(context, name),
+        mlir::StringAttr(), mlir::StringAttr(), mlir::StringAttr(), line, decl);
```
Nit: comments here to spell the default attributes.

abidh replied:
Done.

kiranchandramohan commented on flang/lib/Optimizer/Transforms/AddDebugInfo.cpp:
```diff
@@ -99,6 +109,68 @@ void AddDebugInfoPass::handleDeclareOp(fir::cg::XDeclareOp declOp,
+  scope = getOrCreateModuleAttr(result.second.modules[0], fileAttr, scope,
+                                line - 1, !globalOp.isInitialized());
+
+  auto diType = typeGen.convertType(globalOp.getType(), fileAttr, scope,
```
Nit: Spell the type.

abidh replied:
Done.

kiranchandramohan commented on flang/test/Transforms/debug-module-2.f90:
```diff
@@ -0,0 +1,39 @@
+! RUN: %flang_fc1 -emit-llvm -debug-info-kind=standalone %s -o - | FileCheck %s
+! RUN: %flang_fc1 -emit-llvm -debug-info-kind=line-tables-only %s -o - | FileCheck --check-prefix=LINEONLY %s
```
LLVM IR tests should go to the Integration directory.

abidh replied:
I have moved the testcase to Integration folder. There may be other IR testcase that I added and I will move them in separate PR.

kiranchandramohan commented on flang/test/Transforms/debug-module-1.f90:
```diff
@@ -0,0 +1,39 @@
+! RUN: %flang_fc1 -emit-fir -debug-info-kind=standalone -mmlir --mlir-print-debuginfo %s -o - | \
+! RUN: fir-opt --cg-rewrite --mlir-print-debuginfo | fir-opt --add-debug-info --mlir-print-debuginfo | FileCheck %s
```
Probably best to test the pass only.

abidh replied:
The `AddDebugInfo` pass now operates on `XDeclareOp` so we need to run `CodeGenRewrite` before to convert the `DeclareOp`.

kiranchandramohan replied:
I was trying to convey the following:
-> If it is the `AddDebugInfo` pass that you want to test, then give it the input after `cg-rewrite` and test its output.
-> If it is the `CodeGen` changes that you want to test, then give it the input before `Codegen` and test its output.

I am guessing the LLVM IR changes helps it easier to review and hence it might be OK to add it to the Integration directory. 

Generally tests that specifically checks the changes are easier to review and also maintain. This way of writing helps to ensure that when some unrelated changes happen these tests are not affected. For testing end to end flows we should ideally set up a separate buildbot or move these to the llvm-testsuite/Fortran/Debug.

abidh replied:
I agree with the suggestion on using output of `cg-rewrite` to test the `AddDebugInfo`. My initial motivation for basing test on fortran source code was that it was more readable.

As action items,

I will update `AddDebugInfo` tests to be based on the output of `cg-rewrite`
I will add `codegen` tests when it is involved.
I will keep adding llvm ir tests in Integration for now.

kiranchandramohan reviewed (approved):
LG.

abidh committed:
Handle review comments.

Fixed nits and move IR tests to Integration folder.

abidh committed:
Fix clang-format issues.

abidh committed:
Handle review comments.

Add test files that work on fir and test individual passes. Remove the
fortran testcase.
